crypto/dh/dh_key.c

```diff
@@ -271,6 +271,7 @@ static int generate_key(DH *dh)
             if (dh->params.q == NULL) {
                 /* secret exponent length */
                 l = dh->length ? dh->length : BN_num_bits(dh->params.p) - 1;
+                /* TODO: should l be checked for validity here? */
```

Comment:
I think we need to not be adding anymore TODO's now.

Reply:
In that case, do you have an immediate answer?

Reply:
I think it should do the upper bound check (l < p),
but I wont object to no check.
Maybe add the comment related to PKCS3 here also...

Reply:
Err, this line *is* related to PKCS#3

Reply:
Sigh.. 
I was referring to what you wrote before about PKCS3.....
See PKCS3 7.1 Private-value generation
An integer x, the private value, shall be generated privately and randomly. This integer shall satisfy 0 < x < p−1, unless the central authority specifies a private-value length l, in which case the integer shall satisfy 2l-1 ≤ x < 2l.

Reply:
Ah, gotcha.  Sorry

Reply:
If you don't want to deal with the todo now, remove it and add a issue (which is what was discussed in the last OTC meeting).

Reply:
I think we should just have the validity check.

Reply:
Ok.  Is the bound check that *l* < len(*p*) (where len() is the length in bits) good enough?  Do note that it's stricter than what PKCS#3 stipulates (which is, actually, nothing), albeit logical.

Reply:
Ping?

Reply:
You know, there more I ponder this, the more I think I should just remove the comment.  We might think that PKCS#3 is lacking in this regard, but then so it is.

Reply:
Ah, no, it's me being sloppy...  I forgot about this:

> The central authority may optionally select an integer _l_, the private-value length in bits, that satisfies 2<sup><i>l</i>-1</sup>≤<i>p</i>.

Reply:
Validity check added